### PR TITLE
feat: numeric conditions to fetch legendsets (TECH-788)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Implements [TECH-788](https://jira.dhis2.org/browse/TECH-788)

---

### Key features

1. Fetches available legend sets and legends per legend set for numeric conditions

---

### Description

When a numeric dimension is opened, the user has the option to choose a legend set ("`is one of preset options`"). When this is chosen, all available legend sets for that particular dimension is fetched from the api. When in turn a particular legend set is chosen, all legends for that legend set is fetched from the api as well.

Since there are two endpoints for this, one for data elements (e.g. https://debug.dhis2.org/dev/api/dataElements/qrur9Dvnyt5.json?fields=legendSets%5Bid,name%5D) and one for program attributes (e.g. https://debug.dhis2.org/dev/api/trackedEntityAttributes/B6TnnFMgmCk.json?fields=legendSets[id,name]), the query needs to be switched based on the dimension type. The type metadata itself is fetched from the api when opening a saved AO and saved in the `metadata` prop in the redux store.

---

### Known issues

-   [ ] There's no test data for the `PROGRAM_ATTRIBUTE` dimension type, so this part hasn't been tested  

---

### Screenshots

image

### CODE REVIEW

File: src/components/Dialogs/Conditions/ConditionsManager.js [hunk 1]
```diff
@@ -73,15 +75,15 @@ const ConditionsManager = ({
     onClose,
     setConditionsByDimension,
 }) => {
-    const dimensionType = dimension.valueType
+    const valueType = dimension.valueType
     const isOptionSetCondition =
-        dimensionType === DIMENSION_TYPE_TEXT && dimension.optionSet
+        valueType === DIMENSION_TYPE_TEXT && dimension.optionSet
 
     const [conditionsList, setConditionsList] = useState(
```
Comment (janhenrikoverland) [inline src/components/Dialogs/Conditions/ConditionsManager.js:82]: This piece makes me a bit dizzy..

Comment (janhenrikoverland) [inline src/components/Dialogs/Conditions/ConditionsManager.js:82]: This piece makes me a bit dizzy..

File: src/components/Dialogs/Conditions/ConditionsManager.js [hunk 2]
```diff
@@ -73,15 +75,15 @@ const ConditionsManager = ({
...
         (conditions.condition?.length &&
             parseConditionsStringToArray(conditions.condition)) ||
             (!conditions.condition?.length &&
-                (SINGLETON_TYPES.includes(dimensionType) ||
+                (SINGLETON_TYPES.includes(valueType) ||
```
Comment (janhenrikoverland) [inline src/components/Dialogs/Conditions/ConditionsManager.js:86]: Just wondering why it's called "singleton"

File: src/components/Dialogs/Conditions/NumericCondition.js [hunk 3]
```diff
@@ -249,48 +130,53 @@ const NumericCondition = ({
...
-                        ))}
-                    </SingleSelectField>
-                    {legendSet.legends && (
-                        <MultiSelectField
-                            onChange={({ selected }) =>
-                                setValue(selected.join(';'))
-                            }
-                            inputWidth="330px"
-                            selected={(value?.length && value.split(';')) || []}
+            {operator &&
```
Comment (janhenrikoverland) [inline src/components/Dialogs/Conditions/NumericCondition.js:133]: The first check is not necessary when the second is `operator === OPERATOR_IN`

File: src/components/Dialogs/Conditions/ConditionsManager.js [hunk 4]
```diff
@@ -156,16 +176,21 @@ const ConditionsManager = ({
...
             case DIMENSION_TYPE_PERCENTAGE:
             case DIMENSION_TYPE_INTEGER:
             case DIMENSION_TYPE_INTEGER_POSITIVE:
             case DIMENSION_TYPE_INTEGER_NEGATIVE:
             case DIMENSION_TYPE_INTEGER_ZERO_OR_POSITIVE: {
+                if (!availableLegendSets) {
+                    return null
+                }
+
                 const useDecimalSteps =
```
Comment (janhenrikoverland) [inline src/components/Dialogs/Conditions/ConditionsManager.js:191]: Maybe avoid the hook-like name.

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 2 above)
Comment (janhenrikoverland) [reply src/components/Dialogs/Conditions/ConditionsManager.js:86]: I'm not able to read through this multi-line check and understand what it does.

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 1 above)
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/ConditionsManager.js:82]: Haha how so? Suggestions welcome 🙂  the logic is that option set conditions are `type === TEXT` + has an `optionSet`

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 2 above)
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/ConditionsManager.js:86]: `Singleton` as in _can only be instantiated once_. Some dimensions can have multiple conditions (GT:10;LT:15 etc). They go on separate rows / instances of the corresponding component (e.g. `NumericCondition.js`). However some conditions are single-use, e.g. booleans that will display some checkboxes. You can never display the set of checkboxes multiple times, hence they're "singleton".

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 4 above)
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/ConditionsManager.js]: Changed to `enableDecimalSteps`

File: src/components/Dialogs/Conditions/NumericCondition.js (hunk 3 above)
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/NumericCondition.js]: Good find, removed 👍

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 2 above)
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/ConditionsManager.js:86]: > I'm not able to read through this multi-line check and understand what it does.

No, I agree, it's not super readable. Then again, the logic is quite complex. I could for example extract parts of it separate consts for this purpose only, but I'm not sure if that would make it easier to understand. The key problem is that ConditionsManager handles a lot of different conditions, which themselves vary greatly. Option set, legend set, singleton types and the "regular" types, so it quickly becomes pretty hard to read regardless I think...

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 1 above)
Comment (janhenrikoverland) [reply src/components/Dialogs/Conditions/ConditionsManager.js:82]: I mean the entire block of logic

**Review by janhenrikoverland — APPROVED** [review]
LG, will be thoroughly tested during the next weeks.

**Pull request merged**